File: go/indicators/johnehlers/coronatrendvigor/coronatrendvigor.go
// Package coronatrendvigor implements Ehlers' Corona Trend Vigor heatmap
// indicator.
//
// The Corona trend vigor is computed as the slope of the momentum taken over a
// full dominant cycle period, normalised by the cycle amplitude. The ratio is
// scaled into the range [-10, 10]. A value of ±2 means the trend slope equals
// twice the cycle amplitude; values between -2 and +2 form the "corona" and
// suggest the trend should not be traded.
//
// Reference: John Ehlers, "Measuring Cycle Periods", Stocks & Commodities,
// November 2008.
package coronatrendvigor

//nolint: gofumpt
import (
	"fmt"
	"math"
	"sync"
	"time"

	"zpano/entities"
	"zpano/indicators/core"
	"zpano/indicators/core/outputs"
	"zpano/indicators/johnehlers/corona"
)

const (
	// Bandpass filter delta factor in γ = 1/cos(ω · 2 · 0.1).
	bpDelta = 0.1

	// Ratio EMA coefficients (0.33 * current + 0.67 * previous).
	ratioNewCoef      = 0.33
	ratioPreviousCoef = 0.67

	// Vigor band thresholds.
	vigorMidLow  = 0.3
	vigorMidHigh = 0.7
	vigorMid     = 0.5
	widthEdge    = 0.01

	// Raster update blend factors.
	rasterBlendScale    = 0.8
	rasterBlendPrevious = 0.2
	rasterBlendHalf     = 0.5
	rasterBlendExponent = 0.85

	// Ratio clamp bounds.
	ratioLimit = 10.0

	// vigor = vigorScale*(ratio+ratioLimit); with ratioLimit=10 and range 20,
	// vigorScale=0.05 maps ratio ∈ [-10,10] to vigor ∈ [0,1].
	vigorScale = 0.05
)

// CoronaTrendVigor is Ehlers' Corona Trend Vigor indicator.
type CoronaTrendVigor struct {
	mu                  sync.RWMutex
	mnemonic            string
	description         string
	mnemonicTV          string
	descriptionTV       string
	c                   *corona.Corona
	rasterLength        int
	rasterStep          float64
	maxRasterValue      float64
	minParameterValue   float64
	maxParameterValue   float64
	parameterResolution float64
	raster              []float64
	sampleBuffer        []float64
	sampleCount         int
	samplePrevious      float64
	samplePrevious2     float64
	bandPassPrevious    float64
	bandPassPrevious2   float64
	ratioPrevious       float64
	trendVigor          float64
	barFunc             entities.BarFunc
	quoteFunc           entities.QuoteFunc
	tradeFunc           entities.TradeFunc
}

// NewCoronaTrendVigorDefault returns an instance created with default parameters.
func NewCoronaTrendVigorDefault() (*CoronaTrendVigor, error) {
	return NewCoronaTrendVigorParams(&Params{})
}

// NewCoronaTrendVigorParams returns an instance created with the supplied parameters.
//
//nolint:funlen,cyclop
func NewCoronaTrendVigorParams(p *Params) (*CoronaTrendVigor, error) {
	const (
		invalid      = "invalid corona trend vigor parameters"
		fmtRaster    = "%s: RasterLength should be >= 2"
		fmtMaxRaster = "%s: MaxRasterValue should be > 0"
		fmtMaxParam  = "%s: MaxParameterValue should be > MinParameterValue"
		fmtHP        = "%s: HighPassFilterCutoff should be >= 2"
		fmtMinP      = "%s: MinimalPeriod should be >= 2"
		fmtMaxP      = "%s: MaximalPeriod should be > MinimalPeriod"
		fmtw         = "%s: %w"
		fmtnValue    = "ctv(%d, %g, %g, %g, %d%s)"
		fmtnTV       = "ctv-tv(%d%s)"
		descrValue   = "Corona trend vigor "
		descrTV      = "Corona trend vigor scalar "
		defRaster    = 50
		defMaxRast   = 20.0
		defMinParam  = -10.0
		defMaxParam  = 10.0
		defHPCutoff  = 30
		defMinPer    = 6
		defMaxPer    = 30
	)

	cfg := *p

	if cfg.RasterLength == 0 {
		cfg.RasterLength = defRaster
	}

	if cfg.MaxRasterValue == 0 {
		cfg.MaxRasterValue = defMaxRast
	}

	// MinParameterValue and MaxParameterValue default to -10 and 10; since 0 is a
	// valid user value for either, we only substitute when both are zero (the
	// unconfigured case).
	if cfg.MinParameterValue == 0 && cfg.MaxParameterValue == 0 {
		cfg.MinParameterValue = defMinParam
		cfg.MaxParameterValue = defMaxParam
	}

	if cfg.HighPassFilterCutoff == 0 {
		cfg.HighPassFilterCutoff = defHPCutoff
	}

	if cfg.MinimalPeriod == 0 {
		cfg.MinimalPeriod = defMinPer
	}

	if cfg.MaximalPeriod == 0 {
		cfg.MaximalPeriod = defMaxPer
	}

	if cfg.RasterLength < 2 {
		return nil, fmt.Errorf(fmtRaster, invalid)
	}

	if cfg.MaxRasterValue <= 0 {
		return nil, fmt.Errorf(fmtMaxRaster, invalid)
	}

	if cfg.MaxParameterValue <= cfg.MinParameterValue {
		return nil, fmt.Errorf(fmtMaxParam, invalid)
	}

	if cfg.HighPassFilterCutoff < 2 {
		return nil, fmt.Errorf(fmtHP, invalid)
	}

	if cfg.MinimalPeriod < 2 {
		return nil, fmt.Errorf(fmtMinP, invalid)
	}

	if cfg.MaximalPeriod <= cfg.MinimalPeriod {
		return nil, fmt.Errorf(fmtMaxP, invalid)
	}

	bc := cfg.BarComponent
	if bc == 0 {
		bc = entities.BarMedianPrice
	}

	qc := cfg.QuoteComponent
	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	tc := cfg.TradeComponent
	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	c, err := corona.NewCorona(&corona.Params{
		HighPassFilterCutoff: cfg.HighPassFilterCutoff,
		MinimalPeriod:        cfg.MinimalPeriod,
		MaximalPeriod:        cfg.MaximalPeriod,
	})
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	barFunc, err := entities.BarComponentFunc(bc)
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	quoteFunc, err := entities.QuoteComponentFunc(qc)
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	tradeFunc, err := entities.TradeComponentFunc(tc)
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	componentMnemonic := core.ComponentTripleMnemonic(bc, qc, tc)

	mnemonicValue := fmt.Sprintf(fmtnValue,
		cfg.RasterLength, cfg.MaxRasterValue, cfg.MinParameterValue, cfg.MaxParameterValue,
		cfg.HighPassFilterCutoff, componentMnemonic)
	mnemonicTV := fmt.Sprintf(fmtnTV, cfg.HighPassFilterCutoff, componentMnemonic)

	parameterResolution := float64(cfg.RasterLength-1) / (cfg.MaxParameterValue - cfg.MinParameterValue)

	return &CoronaTrendVigor{
		mnemonic:            mnemonicValue,
		description:         descrValue + mnemonicValue,
		mnemonicTV:          mnemonicTV,
		descriptionTV:       descrTV + mnemonicTV,
		c:                   c,
		rasterLength:        cfg.RasterLength,
		rasterStep:          cfg.MaxRasterValue / float64(cfg.RasterLength),
		maxRasterValue:      cfg.MaxRasterValue,
		minParameterValue:   cfg.MinParameterValue,
		maxParameterValue:   cfg.MaxParameterValue,
		parameterResolution: parameterResolution,
		raster:              make([]float64, cfg.RasterLength),
		sampleBuffer:        make([]float64, c.MaximalPeriodTimesTwo()),
		trendVigor:          math.NaN(),
		barFunc:             barFunc,
		quoteFunc:           quoteFunc,
		tradeFunc:           tradeFunc,
	}, nil
}

// IsPrimed indicates whether the indicator is primed.
func (s *CoronaTrendVigor) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.c.IsPrimed()
}

// Metadata describes the output data of the indicator.
func (s *CoronaTrendVigor) Metadata() core.Metadata {
	return core.BuildMetadata(
		core.CoronaTrendVigor,
		s.mnemonic,
		s.description,
		[]core.OutputText{
			{Mnemonic: s.mnemonic, Description: s.description},
			{Mnemonic: s.mnemonicTV, Description: s.descriptionTV},
		},
	)
}

// Update feeds the next sample and returns the heatmap column plus the current
// TrendVigor. On unprimed bars the heatmap is empty and the scalar is NaN.
//
//nolint:funlen,cyclop,gocognit
func (s *CoronaTrendVigor) Update(sample float64, t time.Time) (*outputs.Heatmap, float64) {
	s.mu.Lock()
	defer s.mu.Unlock()

	if math.IsNaN(sample) {
		return outputs.NewEmptyHeatmap(t, s.minParameterValue, s.maxParameterValue, s.parameterResolution), math.NaN()
	}

	primed := s.c.Update(sample)
	s.sampleCount++

	bufLast := len(s.sampleBuffer) - 1

	if s.sampleCount == 1 {
		s.samplePrevious = sample
		s.sampleBuffer[bufLast] = sample

		return outputs.NewEmptyHeatmap(t, s.minParameterValue, s.maxParameterValue, s.parameterResolution), math.NaN()
	}

	// Bandpass InPhase filter at the dominant cycle median period.
	omega := 2. * math.Pi / s.c.DominantCycleMedian()
	beta2 := math.Cos(omega)
	gamma2 := 1. / math.Cos(omega*2*bpDelta)
	alpha2 := gamma2 - math.Sqrt(gamma2*gamma2-1.)
	bandPass := 0.5*(1-alpha2)*(sample-s.samplePrevious2) +
		beta2*(1+alpha2)*s.bandPassPrevious -
		alpha2*s.bandPassPrevious2

	// Quadrature = derivative / omega.
	quadrature2 := (bandPass - s.bandPassPrevious) / omega

	s.bandPassPrevious2 = s.bandPassPrevious
	s.bandPassPrevious = bandPass
	s.samplePrevious2 = s.samplePrevious
	s.samplePrevious = sample

	// Left-shift sampleBuffer and append the new sample.
	for i := 0; i < bufLast; i++ {
		s.sampleBuffer[i] = s.sampleBuffer[i+1]
	}

	s.sampleBuffer[bufLast] = sample

	// Cycle amplitude.
	amplitude2 := math.Sqrt(bandPass*bandPass + quadrature2*quadrature2)

	// Trend amplitude taken over the cycle period. Use DominantCycleMedian-1
	// directly (the MBST implementation clamps to sampleBuffer length which
	// negates the intent; see the reference impls).
	cyclePeriod := int(s.c.DominantCycleMedian() - 1)
	if cyclePeriod > len(s.sampleBuffer) {
		cyclePeriod = len(s.sampleBuffer)
	}

	if cyclePeriod < 1 {
		cyclePeriod = 1
	}

	lookback := cyclePeriod
	if s.sampleCount < lookback {
		lookback = s.sampleCount
	}

	trend := sample - s.sampleBuffer[len(s.sampleBuffer)-lookback]

	ratio := 0.0
	if math.Abs(trend) > 0 && amplitude2 > 0 {
		ratio = ratioNewCoef*trend/amplitude2 + ratioPreviousCoef*s.ratioPrevious
	}

	if ratio > ratioLimit {
		ratio = ratioLimit
	} else if ratio < -ratioLimit {
		ratio = -ratioLimit
	}

	s.ratioPrevious = ratio

	// ratio ∈ [-10, 10] ⇒ vigor ∈ [0, 1].
	vigor := vigorScale * (ratio + ratioLimit)

	var width float64

	switch {
	case vigor >= vigorMidLow && vigor < vigorMid:
		width = vigor - (vigorMidLow - widthEdge)
	case vigor >= vigorMid && vigor <= vigorMidHigh:
		width = (vigorMidHigh + widthEdge) - vigor
	default:
		width = widthEdge
	}

	s.trendVigor = (s.maxParameterValue-s.minParameterValue)*vigor + s.minParameterValue

	vigorScaledToRasterLength := int(math.Round(float64(s.rasterLength) * vigor))
	vigorScaledToMaxRasterValue := vigor * s.maxRasterValue

	for i := 0; i < s.rasterLength; i++ {
		value := s.raster[i]

		if i == vigorScaledToRasterLength {
			value *= rasterBlendHalf
		} else {
			argument := vigorScaledToMaxRasterValue - s.rasterStep*float64(i)
			if i > vigorScaledToRasterLength {
				argument = -argument
			}

			if width > 0 {
				value = rasterBlendScale *
					(math.Pow(argument/width, rasterBlendExponent) + rasterBlendPrevious*value)
			}
		}

		switch {
		case value < 0:
			value = 0
		case value > s.maxRasterValue, vigor < vigorMidLow, vigor > vigorMidHigh:
			value = s.maxRasterValue
		}

		if math.IsNaN(value) {
			value = 0
		}

		s.raster[i] = value
	}

	if !primed {
		return outputs.NewEmptyHeatmap(t, s.minParameterValue, s.maxParameterValue, s.parameterResolution), math.NaN()
	}

	values := make([]float64, s.rasterLength)
	valueMin := math.Inf(1)
	valueMax := math.Inf(-1)

	for i := 0; i < s.rasterLength; i++ {
		v := s.raster[i]
		values[i] = v

		if v < valueMin {
			valueMin = v
		}

		if v > valueMax {
			valueMax = v
		}
	}

	heatmap := outputs.NewHeatmap(t, s.minParameterValue, s.maxParameterValue, s.parameterResolution,
		valueMin, valueMax, values)

	return heatmap, s.trendVigor
}

// UpdateScalar updates the indicator given the next scalar sample.
func (s *CoronaTrendVigor) UpdateScalar(sample *entities.Scalar) core.Output {
	return s.updateEntity(sample.Time, sample.Value)
}

// UpdateBar updates the indicator given the next bar sample.
func (s *CoronaTrendVigor) UpdateBar(sample *entities.Bar) core.Output {
	return s.updateEntity(sample.Time, s.barFunc(sample))
}

// UpdateQuote updates the indicator given the next quote sample.
func (s *CoronaTrendVigor) UpdateQuote(sample *entities.Quote) core.Output {
	return s.updateEntity(sample.Time, s.quoteFunc(sample))
}

// UpdateTrade updates the indicator given the next trade sample.
func (s *CoronaTrendVigor) UpdateTrade(sample *entities.Trade) core.Output {
	return s.updateEntity(sample.Time, s.tradeFunc(sample))
}

func (s *CoronaTrendVigor) updateEntity(t time.Time, sample float64) core.Output {
	const length = 2

	heatmap, tv := s.Update(sample, t)

	output := make([]any, length)
	output[0] = heatmap
	output[1] = entities.Scalar{Time: t, Value: tv}

	return output
}

File: py/indicators/john_ehlers/corona_trend_vigor/corona_trend_vigor.py
"""Ehlers' Corona Trend Vigor heatmap indicator."""

import math
import datetime

from ....entities.bar import Bar
from ....entities.bar_component import BarComponent, bar_component_value, DEFAULT_BAR_COMPONENT
from ....entities.quote import Quote
from ....entities.quote_component import QuoteComponent, quote_component_value, DEFAULT_QUOTE_COMPONENT
from ....entities.trade import Trade
from ....entities.trade_component import TradeComponent, trade_component_value, DEFAULT_TRADE_COMPONENT
from ....entities.scalar import Scalar
from ...core.identifier import Identifier
from ...core.metadata import Metadata
from ...core.build_metadata import build_metadata, OutputText
from ...core.component_triple_mnemonic import component_triple_mnemonic
from ...core.outputs.heatmap import Heatmap
from ..corona.corona import Corona
from ..corona.params import CoronaParams
from .params import Params

_BP_DELTA = 0.1
_RATIO_NEW_COEF = 0.33
_RATIO_PREVIOUS_COEF = 0.67
_VIGOR_MID_LOW = 0.3
_VIGOR_MID_HIGH = 0.7
_VIGOR_MID = 0.5
_WIDTH_EDGE = 0.01
_RASTER_BLEND_SCALE = 0.8
_RASTER_BLEND_PREVIOUS = 0.2
_RASTER_BLEND_HALF = 0.5
_RASTER_BLEND_EXPONENT = 0.85
_RATIO_LIMIT = 10.0
_VIGOR_SCALE = 0.05


class CoronaTrendVigor:
    """Ehlers' Corona Trend Vigor heatmap indicator."""

    def __init__(self, params: Params) -> None:
        invalid = "invalid corona trend vigor parameters"

        cfg_raster_len = params.raster_length if params.raster_length != 0 else 50
        cfg_max_raster = params.max_raster_value if params.max_raster_value != 0 else 20.0

        if params.min_parameter_value == 0 and params.max_parameter_value == 0:
            cfg_min_param = -10.0
            cfg_max_param = 10.0
        else:
            cfg_min_param = params.min_parameter_value
            cfg_max_param = params.max_parameter_value

        cfg_hp_cutoff = params.high_pass_filter_cutoff if params.high_pass_filter_cutoff != 0 else 30
        cfg_min_period = params.minimal_period if params.minimal_period != 0 else 6
        cfg_max_period = params.maximal_period if params.maximal_period != 0 else 30

        if cfg_raster_len < 2:
            raise ValueError(f"{invalid}: RasterLength should be >= 2")
        if cfg_max_raster <= 0:
            raise ValueError(f"{invalid}: MaxRasterValue should be > 0")
        if cfg_max_param <= cfg_min_param:
            raise ValueError(f"{invalid}: MaxParameterValue should be > MinParameterValue")
        if cfg_hp_cutoff < 2:
            raise ValueError(f"{invalid}: HighPassFilterCutoff should be >= 2")
        if cfg_min_period < 2:
            raise ValueError(f"{invalid}: MinimalPeriod should be >= 2")
        if cfg_max_period <= cfg_min_period:
            raise ValueError(f"{invalid}: MaximalPeriod should be > MinimalPeriod")

        bc = params.bar_component if params.bar_component is not None else BarComponent.MEDIAN
        qc = params.quote_component if params.quote_component is not None else DEFAULT_QUOTE_COMPONENT
        tc = params.trade_component if params.trade_component is not None else DEFAULT_TRADE_COMPONENT

        self._bar_func = bar_component_value(bc)
        self._quote_func = quote_component_value(qc)
        self._trade_func = trade_component_value(tc)

        self._corona = Corona(CoronaParams(
            high_pass_filter_cutoff=cfg_hp_cutoff,
            minimal_period=cfg_min_period,
            maximal_period=cfg_max_period,
        ))

        self._raster_length = cfg_raster_len
        self._raster_step = cfg_max_raster / cfg_raster_len
        self._max_raster_value = cfg_max_raster
        self._min_parameter_value = cfg_min_param
        self._max_parameter_value = cfg_max_param
        self._parameter_resolution = (cfg_raster_len - 1) / (cfg_max_param - cfg_min_param)
        self._raster = [0.0] * cfg_raster_len

        buf_len = self._corona.maximal_period_times_two
        self._sample_buffer = [0.0] * buf_len
        self._sample_count = 0
        self._sample_previous = 0.0
        self._sample_previous2 = 0.0
        self._band_pass_previous = 0.0
        self._band_pass_previous2 = 0.0
        self._ratio_previous = 0.0
        self._trend_vigor = float('nan')

        comp_mn = component_triple_mnemonic(bc, qc, tc)
        self.mnemonic = f"ctv({cfg_raster_len}, {cfg_max_raster:g}, " \
            f"{cfg_min_param:g}, {cfg_max_param:g}, {cfg_hp_cutoff}{comp_mn})"
        self.description = "Corona trend vigor " + self.mnemonic
        self._mnemonic_tv = f"ctv-tv({cfg_hp_cutoff}{comp_mn})"
        self._description_tv = "Corona trend vigor scalar " + self._mnemonic_tv

    def is_primed(self) -> bool:
        return self._corona.is_primed()

    def metadata(self) -> Metadata:
        return build_metadata(
            Identifier.CORONA_TREND_VIGOR,
            self.mnemonic,
            self.description,
            [
                OutputText(mnemonic=self.mnemonic, description=self.description),
                OutputText(mnemonic=self._mnemonic_tv, description=self._description_tv),
            ],
        )

    def update(self, sample: float, t: datetime.datetime) -> tuple:
        """Returns (heatmap, trend_vigor)."""
        if math.isnan(sample):
            return (Heatmap.empty(t, self._min_parameter_value,
                                  self._max_parameter_value, self._parameter_resolution),
                    float('nan'))

        primed = self._corona.update(sample)
        self._sample_count += 1

        buf_last = len(self._sample_buffer) - 1

        if self._sample_count == 1:
            self._sample_previous = sample
            self._sample_buffer[buf_last] = sample
            return (Heatmap.empty(t, self._min_parameter_value,
                                  self._max_parameter_value, self._parameter_resolution),
                    float('nan'))

        # Bandpass filter at the dominant cycle median period.
        dcm = self._corona.dominant_cycle_median
        omega = 2.0 * math.pi / dcm if dcm != 0 else math.inf
        beta2 = math.cos(omega)
        cos_val = math.cos(omega * 2 * _BP_DELTA)
        if cos_val == 0:
            gamma2 = math.copysign(math.inf, 1.0)
        else:
            gamma2 = 1.0 / cos_val
        disc = gamma2 * gamma2 - 1.0
        alpha2 = gamma2 - math.sqrt(disc) if disc >= 0 else 0.0
        band_pass = 0.5 * (1 - alpha2) * (sample - self._sample_previous2) + \
            beta2 * (1 + alpha2) * self._band_pass_previous - \
            alpha2 * self._band_pass_previous2

        # Quadrature = derivative / omega.
        if omega == 0:
            diff = band_pass - self._band_pass_previous
            quadrature2 = math.copysign(math.inf, diff) if diff != 0 else 0.0
        else:
            quadrature2 = (band_pass - self._band_pass_previous) / omega

        self._band_pass_previous2 = self._band_pass_previous
        self._band_pass_previous = band_pass
        self._sample_previous2 = self._sample_previous
        self._sample_previous = sample

        # Left-shift sample buffer and append new sample.
        for i in range(buf_last):
            self._sample_buffer[i] = self._sample_buffer[i + 1]
        self._sample_buffer[buf_last] = sample

        # Cycle amplitude.
        amplitude2 = math.sqrt(band_pass * band_pass + quadrature2 * quadrature2)

        # Trend over cycle period.
        dcm = self._corona.dominant_cycle_median
        if math.isinf(dcm) or math.isnan(dcm):
            cycle_period = len(self._sample_buffer)
        else:
            cycle_period = int(dcm - 1)
            if cycle_period > len(self._sample_buffer):
                cycle_period = len(self._sample_buffer)
        if cycle_period < 1:
            cycle_period = 1

        lookback = cycle_period
        if self._sample_count < lookback:
            lookback = self._sample_count

        trend = sample - self._sample_buffer[len(self._sample_buffer) - lookback]

        ratio = 0.0
        if abs(trend) > 0 and amplitude2 > 0:
            ratio = _RATIO_NEW_COEF * trend / amplitude2 + _RATIO_PREVIOUS_COEF * self._ratio_previous

        if ratio > _RATIO_LIMIT:
            ratio = _RATIO_LIMIT
        elif ratio < -_RATIO_LIMIT:
            ratio = -_RATIO_LIMIT

        self._ratio_previous = ratio

        # vigor in [0, 1]
        vigor = _VIGOR_SCALE * (ratio + _RATIO_LIMIT)

        # Width.
        if vigor >= _VIGOR_MID_LOW and vigor < _VIGOR_MID:
            width = vigor - (_VIGOR_MID_LOW - _WIDTH_EDGE)
        elif vigor >= _VIGOR_MID and vigor <= _VIGOR_MID_HIGH:
            width = (_VIGOR_MID_HIGH + _WIDTH_EDGE) - vigor
        else:
            width = _WIDTH_EDGE

        self._trend_vigor = (self._max_parameter_value - self._min_parameter_value) * \
            vigor + self._min_parameter_value

        vigor_scaled_to_raster_length = round(self._raster_length * vigor)
        vigor_scaled_to_max_raster_value = vigor * self._max_raster_value

        for i in range(self._raster_length):
            value = self._raster[i]

            if i == vigor_scaled_to_raster_length:
                value *= _RASTER_BLEND_HALF
            else:
                argument = vigor_scaled_to_max_raster_value - self._raster_step * i
                if i > vigor_scaled_to_raster_length:
                    argument = -argument

                if width > 0:
                    value = _RASTER_BLEND_SCALE * \
                        (math.pow(argument / width, _RASTER_BLEND_EXPONENT) + _RASTER_BLEND_PREVIOUS * value)

            if value < 0:
                value = 0
            elif value > self._max_raster_value or vigor < _VIGOR_MID_LOW or vigor > _VIGOR_MID_HIGH:
                value = self._max_raster_value

            if math.isnan(value):
                value = 0

            self._raster[i] = value

        if not primed:
            return (Heatmap.empty(t, self._min_parameter_value,
                                  self._max_parameter_value, self._parameter_resolution),
                    float('nan'))

        values = list(self._raster)
        value_min = min(values)
        value_max = max(values)

        heatmap = Heatmap(t, self._min_parameter_value, self._max_parameter_value,
                          self._parameter_resolution, value_min, value_max, values)

        return heatmap, self._trend_vigor

    def update_scalar(self, sample: Scalar) -> list:
        return self._update_entity(sample.time, sample.value)

    def update_bar(self, sample: Bar) -> list:
        return self._update_entity(sample.time, self._bar_func(sample))

    def update_quote(self, sample: Quote) -> list:
        return self._update_entity(sample.time, self._quote_func(sample))

    def update_trade(self, sample: Trade) -> list:
        return self._update_entity(sample.time, self._trade_func(sample))

    def _update_entity(self, t: datetime.datetime, sample: float) -> list:
        heatmap, tv = self.update(sample, t)
        return [heatmap, Scalar(t, tv)]
